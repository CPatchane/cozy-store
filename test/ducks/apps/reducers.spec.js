'use strict'

/* eslint-env jest */

import _get from 'lodash.get'

import {
  LOADING_APP,
  LOADING_APP_INTENT,
  FETCH_APPS,
  FETCH_APPS_SUCCESS,
  FETCH_APPS_FAILURE,
  FETCH_APP,
  FETCH_APP_SUCCESS,
  FETCH_APP_FAILURE,
  FETCH_REGISTRY_APPS_SUCCESS,
  UNINSTALL_APP,
  UNINSTALL_APP_SUCCESS,
  UNINSTALL_APP_FAILURE,
  INSTALL_APP,
  INSTALL_APP_SUCCESS,
  INSTALL_APP_FAILURE,
  list,
  isFetching,
  isAppFetching,
  isInstalling,
  isUninstalling,
  actionError,
  fetchError
} from 'ducks/apps/reducers'
import mockApps from './_mockApps'
import mockApp from './_mockPhotosRegistryVersion'

const mockError = new Error('This is a test error')

const mockRegistryApps = mockApps.filter(a => a.isInRegistry)
const mockAppAlone = [mockApp]

const reducersMap = new Map([
  ['list', list],
  ['isFetching', isFetching],
  ['isAppFetching', isAppFetching],
  ['isInstalling', isInstalling],
  ['isUninstalling', isUninstalling],
  ['actionError', actionError],
  ['fetchError', fetchError]
])

const actionsMap = new Map([
  [
    'FETCH_APPS*',
    {
      fetchAppsAction: { type: FETCH_APPS },
      fetchAppsSuccessAction: {
        apps: mockApps,
        type: FETCH_APPS_SUCCESS
      },
      fetchAppsErrorAction: {
        error: mockError,
        type: FETCH_APPS_FAILURE
      }
    }
  ],
  [
    'FETCH_REGISTRY_APPS*',
    {
      fetchRegistryAppsSuccessAction: {
        apps: mockRegistryApps,
        type: FETCH_REGISTRY_APPS_SUCCESS
      }
    }
  ],
  [
    'FETCH_APP*',
    {
      fetchAppAction: { type: FETCH_APP },
      fetchAppSuccessAction: {
        apps: mockAppAlone,
        type: FETCH_APP_SUCCESS
      },
      fetchAppErrorAction: {
        error: mockError,
        type: FETCH_APP_FAILURE
      }
    }
  ],
  [
    'INSTALL_APP*',
    {
      installAppAction: { type: INSTALL_APP },
      installAppSuccessAction: {
        installedApp: mockApp,
        type: INSTALL_APP_SUCCESS
      },
      installAppErrorAction: {
        error: mockError,
        type: INSTALL_APP_FAILURE
      }
    }
  ],
  [
    'UNINSTALL_APP*',
    {
      uninstallAppAction: { type: UNINSTALL_APP },
      uninstallAppSuccessAction: {
        slug: 'collect',
        type: UNINSTALL_APP_SUCCESS
      },
      uninstallAppErrorAction: {
        error: mockError,
        type: UNINSTALL_APP_FAILURE
      }
    }
  ],
  [
    'LOADING_APP*',
    {
      loadingAppAction: { type: LOADING_APP },
      loadingAppIntentAction: { type: LOADING_APP_INTENT }
    }
  ]
])

/*
  each configuration line -> actionName: [input, assertPath, expectedResult]
  Missing action name property means that action doesn't touch the state
  for this reducer.
*/
const reducersTestConfig = {
  list: {
    installAppSuccessAction: [
      mockApps.map(a => {
        if (
          a.slug ==
          actionsMap.get('INSTALL_APP*').installAppSuccessAction.installedApp
            .slug
        )
          a.installed = false
        return a
      }),
      'toEqual',
      mockApps.map(a => {
        if (
          a.slug ==
          actionsMap.get('INSTALL_APP*').installAppSuccessAction.installedApp
            .slug
        )
          a.installed = true
        return a
      })
    ],
    uninstallAppSuccessAction: [
      mockApps.map(a => {
        if (
          a.slug ==
          actionsMap.get('UNINSTALL_APP*').uninstallAppSuccessAction.slug
        )
          a.installed = true
        return a
      }),
      'toEqual',
      mockApps.map(a => {
        if (
          a.slug ==
          actionsMap.get('UNINSTALL_APP*').uninstallAppSuccessAction.slug
        )
          a.installed = false
        return a
      })
    ],
    fetchAppsSuccessAction: [[], 'toEqual', mockApps],
    fetchRegistryAppsSuccessAction: [[], 'toEqual', mockRegistryApps],
    fetchAppSuccessAction: [[], 'toEqual', mockAppAlone]
  },
  isFetching: {
    loadingAppAction: [false, 'toBe', true],
    fetchAppsAction: [false, 'toBe', true],
    fetchAppsErrorAction: [true, 'toBe', false],
    fetchAppsSuccessAction: [true, 'toBe', false]
  },
  isAppFetching: {
    loadingAppIntentAction: [false, 'toBe', true],
    fetchAppAction: [false, 'toBe', true],
    fetchAppErrorAction: [true, 'toBe', false],
    fetchAppSuccessAction: [true, 'toBe', false]
  },
  isInstalling: {
    installAppAction: [false, 'toBe', true],
    installAppSuccessAction: [true, 'toBe', false],
    installAppErrorAction: [true, 'toBe', false]
  },
  isUninstalling: {
    uninstallAppAction: [false, 'toBe', true],
    uninstallAppSuccessAction: [true, 'toBe', false],
    uninstallAppErrorAction: [true, 'toBe', false]
  },
  actionError: {
    installAppSuccessAction: [mockError, 'toBe', null],
    installAppErrorAction: [null, 'toBe', mockError],
    uninstallAppSuccessAction: [mockError, 'toBe', null],
    uninstallAppErrorAction: [null, 'toBe', mockError]
  },
  fetchError: {
    fetchAppsSuccessAction: [mockError, 'toBe', null],
    fetchAppsErrorAction: [null, 'toBe', mockError],
    fetchAppSuccessAction: [mockError, 'toBe', null],
    fetchAppErrorAction: [null, 'toBe', mockError]
  }
}

describe('Apps ducks reducers', () => {
  function expectToNotTouchTheState(reducer, action) {
    expect(reducer('default_state', action)).toBe('default_state')
  }
  Object.keys(reducersTestConfig).map(reducerName => {
    // for each reducer from the config
    if (reducersTestConfig.hasOwnProperty(reducerName)) {
      describe(reducerName, () => {
        const reducerConfig = reducersTestConfig[reducerName]
        const reducer = reducersMap.get(reducerName)
        // we get all actions group defined in the actions Map
        actionsMap.forEach((actionsGroup, actionsGroupName) => {
          it(`- ${actionsGroupName}`, () => {
            beforeEach(() => {
              jest.resetAllMocks()
              jest.resetModules()
            })
            // we test all actions from the group
            Object.keys(actionsGroup).map(actionName => {
              const action = actionsGroup[actionName]
              // if defined in the reducer config, we use provided parameters
              if (reducerConfig.hasOwnProperty(actionName)) {
                const params = reducerConfig[actionName]
                _get(expect(reducer(params[0], action)), params[1])(params[2])
              } else {
                // not defined in the config, we guess that this is action
                // which doesn't change the reducer state here
                expectToNotTouchTheState(reducer, action)
              }
            })
          })
        })
      })
    }
  })
})
