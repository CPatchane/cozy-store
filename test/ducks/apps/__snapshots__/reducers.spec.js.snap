// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Apps ducks helpers _consolidateApps should merge correctly two apps infos lists 1`] = `
Array [
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "installed": false,
    "isInRegistry": true,
    "name": "Collect",
    "name_prefix": "Cozy",
    "permissions": Object {
      "mock": Object {
        "type": "io.mock.doctype",
      },
      "mock2": Object {
        "type": "io.mock.doctype2",
      },
    },
    "related": "http://collect.cozy.mock/",
    "slug": "collect",
    "tags": Array [
      "konnector",
      "collect",
      "bills",
      "providers",
      "files",
    ],
    "type": "webapp",
    "uninstallable": false,
    "version": "3.0.0",
    "versions": Object {
      "beta": Array [
        "3.0.0",
      ],
      "dev": Array [
        "3.0.0",
      ],
      "stable": Array [
        "3.0.0",
      ],
    },
  },
  Object {
    "categories": Array [
      "others",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "isInRegistry": true,
    "locales": Object {
      "name": "For Dev Only",
    },
    "name": "DevOnly",
    "slug": "devonly",
    "type": "webapp",
    "uninstallable": true,
    "versions": Object {
      "beta": Array [],
      "dev": Array [
        "1.0.0-betaojdkehy989ekhflldh",
      ],
      "stable": Array [],
    },
  },
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "installed": true,
    "name": "Drive",
    "name_prefix": "Cozy",
    "permissions": Object {
      "mock2": Object {
        "type": "io.mock.doctype2",
      },
    },
    "related": "http://drive.cozy.mock/",
    "slug": "drive",
    "tags": Array [
      "search",
      "files",
      "folders",
    ],
    "type": "webapp",
    "uninstallable": false,
    "version": "3.0.0-beta89bnhj3993",
  },
  Object {
    "categories": Array [
      "isp",
      "telecom",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "icon": "<svg></svg>",
    "isInRegistry": true,
    "name": "Bouilligue",
    "slug": "konnector-bouilligue",
    "type": "konnector",
    "uninstallable": true,
    "versions": Object {
      "beta": Array [
        "0.1.0",
      ],
      "dev": Array [
        "0.1.0",
      ],
      "stable": Array [
        "0.1.0",
      ],
    },
  },
  Object {
    "categories": Array [
      "transport",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "icon": "<svg></svg>",
    "installed": true,
    "isInRegistry": true,
    "name": "Trinlane",
    "permissions": Object {
      "mock": Object {
        "type": "io.mock.doctype",
      },
    },
    "slug": "konnector-trinlane",
    "tags": Array [
      "transport",
      "files",
      "bills",
    ],
    "type": "konnector",
    "uninstallable": true,
    "version": "0.1.0",
    "versions": Object {
      "beta": Array [
        "0.1.0",
      ],
      "dev": Array [
        "0.1.0",
      ],
      "stable": Array [
        "0.1.0",
      ],
    },
  },
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
      "url": "https://cozy.io",
    },
    "editor": "Cozy",
    "icon": "https://mockcozy.cc/registry/photos/3.0.0/icon.svg",
    "installed": true,
    "isInRegistry": true,
    "langs": Array [
      "en",
      "fr",
    ],
    "licence": "AGPL-3.0",
    "locales": Object {
      "en": Object {
        "changes": "### NEW!!!

 Now on the cozy-store :tada",
        "long_description": "### Features

Enjoy features to grab your data in you Cozy :tada:",
        "short_description": "Gestionnaire de photos pour Cozy v3",
      },
      "fr": Object {
        "long_description": "### Photos

Gestionnaire de photos pour Cozy v3 :tada:",
        "short_description": "Gestionnaire de photos pour Cozy v3",
      },
    },
    "name": "Photos",
    "name_prefix": "Cozy",
    "permissions": Object {
      "albums": Object {
        "description": "Required to manage photos albums",
        "methods": Array [
          "GET",
          "POST",
          "PUT",
        ],
        "type": "io.cozy.photos.albums",
      },
      "apps": Object {
        "description": "Required by the cozy-bar to display the icons of the apps",
        "type": "io.cozy.apps",
        "verbs": Array [
          "GET",
        ],
      },
      "contacts": Object {
        "description": "Required to to share photos with your contacts",
        "methods": Array [
          "GET",
          "POST",
        ],
        "type": "io.cozy.contacts",
      },
      "files": Object {
        "description": "Required for photo access",
        "methods": Array [
          "GET",
          "POST",
          "PUT",
        ],
        "type": "io.cozy.files",
      },
      "settings": Object {
        "description": "Required by the cozy-bar to display Claudy and know which applications are coming soon",
        "type": "io.cozy.settings",
        "verbs": Array [
          "GET",
        ],
      },
    },
    "platforms": Array [
      Object {
        "type": "ios",
        "url": "",
      },
      Object {
        "type": "linux",
        "url": "",
      },
      Object {
        "type": "android",
        "url": "https://mock.app",
      },
    ],
    "routes": Object {
      "/": Object {
        "folder": "/",
        "index": "index.html",
        "public": false,
      },
      "/public": Object {
        "folder": "/public",
        "index": "index.html",
        "public": true,
      },
    },
    "short_description": "Photos manager for Cozy v3",
    "slug": "photos",
    "source": "https://github.com/cozy/cozy-drive.git@build-photos",
    "type": "webapp",
    "uninstallable": true,
    "version": "3.0.0",
    "versions": Object {
      "beta": Array [
        "3.0.0",
      ],
      "dev": Array [
        "3.0.0",
      ],
      "stable": Array [
        "3.0.0",
      ],
    },
  },
  Object {
    "availableVersion": true,
    "categories": Array [
      "partners",
    ],
    "icon": "<svg></svg>",
    "installed": true,
    "isInRegistry": true,
    "name": "Tasky",
    "slug": "tasky",
    "type": "webapp",
    "uninstallable": true,
    "versions": Object {
      "beta": Array [
        "1.0.0",
      ],
      "dev": Array [
        "1.0.0",
      ],
      "stable": Array [
        "1.0.0",
      ],
    },
  },
]
`;

exports[`Apps ducks helpers _sortAlphabetically should sort alphabetically according to the provided property name 1`] = `
Array [
  Object {
    "slug": "aya",
  },
  Object {
    "slug": "bab",
  },
  Object {
    "slug": "bibi",
  },
]
`;

exports[`Apps ducks helpers _sortAlphabetically should sort alphabetically according to the provided property name 2`] = `
Array [
  Object {
    "custom": "aya",
  },
  Object {
    "custom": "bab",
  },
  Object {
    "custom": "bibi",
  },
]
`;
